partiql-parser/src/result.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright Amazon.com, Inc. or its affiliates.
+
+//! [`Error`] and [`Result`] types for parsing PartiQL.
+
+use pest::error::{ErrorVariant, LineColLocation};
+use std::fmt;
+use thiserror::Error;
+
+/// Position in the source for an error.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum Position {
+    Unknown,
+    At { line: usize, column: usize },
+}
+
+impl fmt::Display for Position {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Position::Unknown => write!(f, "UNKNOWN POSITION"),
+            Position::At { line, column } => write!(f, "line {}, column {}", *line, *column),
+        }
+    }
+}
+
+/// Errors from the PartiQL parser.
+#[derive(Clone, Debug, Eq, PartialEq, Error)]
+pub enum ParserError {
+    #[error("Syntax Error: {message} ({position})")]
+    SyntaxError { message: String, position: Position },
+}
+
+impl ParserError {
+    /// Convenience function to create a [SyntaxError].
+    fn syntax_error<S: Into<String>>(message: S, position: Position) -> Self {
+        Self::SyntaxError {
+            message: message.into(),
+            position,
+        }
+    }
+}
+
+impl<R> From<pest::error::Error<R>> for ParserError
+where
+    R: fmt::Debug,
+{
+    fn from(error: pest::error::Error<R>) -> Self {
```

Comment:
This makes it so we can convert a Pest error to a `ParserError` automatically via the `?` error handling operator in Rust.